homeassistant/helpers/icon.py

```diff
@@ -1,7 +1,188 @@
 """Icon helper methods."""
 from __future__ import annotations
 
+import asyncio
+from collections.abc import Iterable
 from functools import lru_cache
+import logging
+from typing import Any
+
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.loader import Integration, async_get_integrations
+from homeassistant.util.json import load_json
+
+from .translation import build_resources, recursive_flatten
+
+ICON_LOAD_LOCK = "icon_load_lock"
+ICON_FLATTEN_CACHE = "icon_flatten_cache"
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@callback
+def component_icons_path(component: str, integration: Integration) -> str | None:
+    """Return the icons json file location for a component.
+
+    Ex: components/hue/icons.json
+    If component is just a single file, will return None.
+    """
+    domain = component.rpartition(".")[-1]
+
+    # If it's a component that is just one file, we don't support icons
+    # Example custom_components/my_component.py
+    if integration.file_path.name != domain:
+        return None
+
+    return str(integration.file_path / "icons.json")
+
+
+def load_icons_files(icons_files: dict[str, str]) -> dict[str, dict[str, Any]]:
+    """Load and parse icons.json files."""
+    loaded = {}
+    for component, icons_file in icons_files.items():
+        loaded_json = load_json(icons_file)
+
+        if not isinstance(loaded_json, dict):
+            _LOGGER.warning(
+                "Icons file is unexpected type %s. Expected dict for %s",
+                type(loaded_json),
+                icons_file,
+            )
+            continue
+
+        loaded[component] = loaded_json
+
+    return loaded
+
+
+async def _async_get_component_icons(
+    hass: HomeAssistant,
+    components: set[str],
+    integrations: dict[str, Integration],
+) -> dict[str, Any]:
+    """Load icons."""
+    icons: dict[str, Any] = {}
+
+    # Determine files to load
+    files_to_load = {}
+    for loaded in components:
+        domain = loaded.rpartition(".")[-1]
+        if (path := component_icons_path(loaded, integrations[domain])) is None:
+            icons[loaded] = {}
+        else:
+            files_to_load[loaded] = path
+
+    if not files_to_load:
+        return icons
+
+    # Load files
+    if load_icons_job := hass.async_add_executor_job(load_icons_files, files_to_load):
+        icons |= await load_icons_job
+
+    return icons
+
+
+class _IconsCache:
+    """Cache for flattened icons."""
+
+    __slots__ = ("hass", "loaded", "cache")
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize the cache."""
+        self.hass = hass
```

Comment:
Even if the class is currently protected, we should still mark variables that should not be accessed from outside the class as protected.